examples/hpc/dla-driver/src/lib.rs

```diff
@@ -18,291 +100,493 @@ macro_rules! set_bits {
     }
-}
 
-pub fn dla_set_input_data_bank(bank: usize) {
-    let mut reg = dla_read_reg(DLA_BUF_DATA_BANK);
-    reg = set_bits!(
-        DLA_BUF_DATA_BANK_B_OFFSET,
-        DLA_BUF_DATA_BANK_B_BITMASK,
-        reg,
-        bank
-    );
-    dla_write_reg(DLA_BUF_DATA_BANK, reg);
-}
+    pub fn read_reg(&self, offset: usize) -> u32 {
+        unsafe { ptr::read_volatile((DLA0_ADDR + offset) as *mut u32) }
+    }
 
-pub fn dla_set_kernel_data_bank(bank: usize) {
-    let mut reg = dla_read_reg(DLA_BUF_DATA_BANK);
-    reg = set_bits!(
-        DLA_BUF_DATA_BANK_A_OFFSET,
-        DLA_BUF_DATA_BANK_A_BITMASK,
-        reg,
-        bank
-    );
-    dla_write_reg(DLA_BUF_DATA_BANK, reg);
-}
+    pub fn read(&self, buf: &mut [u8], len: usize, offset: usize) {
+        for i in 0..len {
+            unsafe { buf[i] = ptr::read_volatile((DLA0_ADDR + offset + i) as *mut u8) }
+        }
+    }
 
-pub fn dla_set_input_size(channels: usize, width: usize, height: usize) {
-    let mut reg = 0;
-    reg = set_bits!(
-        DLA_BUF_INPUT_CHANNELS_OFFSET,
-        DLA_BUF_INPUT_CHANNELS_BITMASK,
-        reg,
-        channels - 1
-    );
-    reg = set_bits!(
-        DLA_BUF_INPUT_WIDTH_OFFSET,
-        DLA_BUF_INPUT_WIDTH_BITMASK,
-        reg,
-        width - 1
-    );
-    reg = set_bits!(
-        DLA_BUF_INPUT_HEIGHT_OFFSET,
-        DLA_BUF_INPUT_HEIGHT_BITMASK,
-        reg,
-        height - 1
-    );
-    dla_write_reg(DLA_BUF_INPUT, reg);
-}
+    pub fn write_data_bank(&self, offset: usize, buf: &mut [u8]) {
+        //sprintln!("\nWrite to bank {:#x}, data: {:?}", offset, buf);
+        for (i, b) in buf.iter().enumerate() {
+            unsafe { ptr::write_volatile((MEMORY_BANK_BASE_ADDR + offset + i) as *mut u8, *b) };
+        }
+    }
 
-pub fn dla_set_kernel_size(channels: usize, width: usize, height: usize) {
-    let mut reg = 0;
-    reg = set_bits!(
-        DLA_BUF_KERNEL_0_S_CHANNELS_OFFSET,
-        DLA_BUF_KERNEL_0_S_CHANNELS_BITMASK,
-        reg,
-        channels - 1
-    );
-    reg = set_bits!(
-        DLA_BUF_KERNEL_0_WIDTH_OFFSET,
-        DLA_BUF_KERNEL_0_WIDTH_BITMASK,
-        reg,
-        width - 1
-    );
-    reg = set_bits!(
-        DLA_BUF_KERNEL_0_HEIGHT_OFFSET,
-        DLA_BUF_KERNEL_0_HEIGHT_BITMASK,
-        reg,
-        height - 1
-    );
-    dla_write_reg(DLA_BUF_KERNEL_0, reg);
-}
+    pub fn read_data_bank_offset(&self, offset: usize) -> u128 {
+        // NOTE: this function enforces the 128-bit addressing
+        if cfg!(feature = "vp") {
+            let mut result: u128 = 0;
+            for i in 0..4 {
+                result |= (unsafe {
+                    ptr::read_volatile((MEMORY_BANK_BASE_ADDR + offset + (i * 4)) as *mut u32)
+                } as u128)
+                    << (32 * i)
+            }
+            result
+        } else {
+            unsafe { ptr::read_volatile((MEMORY_BANK_BASE_ADDR + (offset & !0xF)) as *mut u128) }
```

Comment:
What's going on here with the mystery 8-bit bitmask? :)

Reply:
The bitmask makes sure that the buffer access is aligned to 128-bits.

Reply:
This is sus. I'd delegate this to a user problem. Just assert the alignment instead of editing the input.

Reply:
Removed